src/coreclr/src/vm/classlayoutinfo.cpp

```diff
@@ -0,0 +1,1105 @@
+    LoaderAllocator* pAllocator = pMT->GetLoaderAllocator();
+    AllocMemHolder<EEClassNativeLayoutInfo> pNativeLayoutInfo(
+        pAllocator->GetLowFrequencyHeap()->AllocMem(
+            S_SIZE_T(sizeof(EEClassNativeLayoutInfo)) + S_SIZE_T(sizeof(NativeFieldDescriptor)) * S_SIZE_T(numTotalInstanceFields)));
+
+    pNativeLayoutInfo->m_numFields = numTotalInstanceFields;
+
+    // Now compute the native size of each field
+    for (LayoutRawFieldInfo* pfwalk = pInfoArray; pfwalk->m_MD != mdFieldDefNil; pfwalk++)
+    {
+        pfwalk->m_placement.m_size = pfwalk->m_nfd.NativeSize();
+        pfwalk->m_placement.m_alignment = pfwalk->m_nfd.AlignmentRequirement();
+    }
+
+    S_UINT32 cbSortArraySize = S_UINT32(cInstanceFields) * S_UINT32(sizeof(LayoutRawFieldInfo*));
+    if (cbSortArraySize.IsOverflow())
+    {
+        ThrowHR(COR_E_TYPELOAD);
+    }
+
+    BOOL fExplicitOffsets = pMT->GetClass()->HasExplicitFieldOffsetLayout();
+
+    LayoutRawFieldInfo** pSortArray = (LayoutRawFieldInfo**)_alloca(cbSortArraySize.Value());
+    SetOffsetsAndSortFields(pInternalImport, pMT->GetCl(), pInfoArray, cInstanceFields, fExplicitOffsets, cbAdjustedParentLayoutNativeSize, pModule, pSortArray);
+
+    EEClassLayoutInfo* pEEClassLayoutInfo = pMT->GetLayoutInfo();
+
+    ULONG classSizeInMetadata = 0;
+    if (pEEClassLayoutInfo->HasExplicitSize())
+    {
+        HRESULT hr = pInternalImport->GetClassTotalSize(pMT->GetCl(), &classSizeInMetadata);
+
+        CONSISTENCY_CHECK(hr == S_OK);
+    }
+
+    BYTE parentAlignmentRequirement = 0;
+    if (fParentHasLayout)
+    {
+        parentAlignmentRequirement = pParentLayoutInfo->GetLargestAlignmentRequirement();
+    }
+
+    CalculateSizeAndFieldOffsets(
+        cbAdjustedParentLayoutNativeSize,
+        cInstanceFields,
+        fExplicitOffsets,
+        pSortArray,
+        classSizeInMetadata,
+        pMT->GetLayoutInfo()->GetPackingSize(),
+        parentAlignmentRequirement,
+        /*limitToMaxInteropSize*/ TRUE,
+        &pNativeLayoutInfo->m_alignmentRequirement,
+        &pNativeLayoutInfo->m_size);
+
+    if (pNativeLayoutInfo->m_size == 0)
+    {
+        _ASSERTE(pEEClassLayoutInfo->IsZeroSized());
+        pNativeLayoutInfo->m_size = 1; // Bump the managed size of the structure up to 1.
+    }
+
+    // The intrinsic Vector<T> type has a special size. Copy the native size and alignment
+    // from the managed size and alignment.
+
+    if (pModule->IsSystem() && strcmp(szName, "Vector`1") == 0 && strcmp(szNamespace, "System.Numerics") == 0)
+    {
+        pNativeLayoutInfo->m_size = pEEClassLayoutInfo->GetManagedSize();
+        pNativeLayoutInfo->m_alignmentRequirement = pEEClassLayoutInfo->m_ManagedLargestAlignmentRequirementOfAllMembers;
+    }
+
+    // The intrinsic Vector types have specialized alignment requirements. For these types,
+    // copy the managed alignment requirement as the native alignment requirement.
+    if (strcmp(szNamespace, g_IntrinsicsNS) == 0)
```

Comment:
The same CoreLib check should be here. Since you have MethodTable available, you can just check for `MethodTable::IsIntrinsicType()` for both.

Also, you do not need to fetch the name for these checks. You can check for them using `pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__VECTOR64T)`, etc. like the marshaler generator.